File: android/app/src/androidTest/java/host/exp/exponent/utils/ExponentMatchers.java
// Copyright 2015-present 650 Industries. All rights reserved.

package host.exp.exponent.utils;

import android.support.test.espresso.matcher.BoundedMatcher;
import android.view.View;

import org.hamcrest.Description;
import org.hamcrest.Matcher;

import static org.hamcrest.Matchers.is;

public class ExponentMatchers {

  public static String getTestId(View view) {
    return view.getTag() instanceof String ? (String) view.getTag() : null;
  }

  public static Matcher<View> withTestId(String text) {
    return withTestId(is(text));
  }

  public static Matcher<View> withTestId(final Matcher<String> stringMatcher) {
    return new BoundedMatcher<View, View>(View.class) {
      @Override
      public void describeTo(Description description) {
        description.appendText("with test id: ");
        stringMatcher.describeTo(description);
      }

      @Override
      public boolean matchesSafely(View view) {
        String testId = getTestId(view);
        if (testId == null) {
          return false;
        } else {
          return stringMatcher.matches(testId);
        }
      }
    };
  }
}

File: apps/expo-go/android/app/src/androidTest/java/host/exp/exponent/utils/ExponentMatchers.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package host.exp.exponent.utils

import android.view.View
import androidx.test.espresso.matcher.BoundedMatcher
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.Matchers

object ExponentMatchers {
  fun getTestId(view: View): String? {
    return if (view.tag is String) view.tag as String else null
  }

  fun withTestId(text: String): Matcher<View> {
    return withTestId(Matchers.`is`(text))
  }

  private fun withTestId(stringMatcher: Matcher<String>): Matcher<View> {
    return object : BoundedMatcher<View, View>(View::class.java) {
      override fun describeTo(description: Description) {
        description.appendText("with test id: ")
        stringMatcher.describeTo(description)
      }

      public override fun matchesSafely(view: View): Boolean {
        val testId = getTestId(view)
        return if (testId == null) {
          false
        } else {
          stringMatcher.matches(testId)
        }
      }
    }
  }
}
